src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -660,26 +658,60 @@ ImmutableArray<byte> computeContentHash()
             }
         }
 
-        internal static ImmutableArray<byte> CalculateChecksum(byte[] buffer, int offset, int count, SourceHashAlgorithm algorithmId)
+#if NETCOREAPP
+        internal static ImmutableArray<byte> CalculateChecksum(ReadOnlySpan<byte> buffer, SourceHashAlgorithm algorithmId)
         {
-            using (var algorithm = CryptographicHashProvider.TryGetAlgorithm(algorithmId))
+            using var algorithm = CryptographicHashProvider.GetAlgorithm(algorithmId);
+            int size = algorithmId switch
+            {
+                SourceHashAlgorithm.Sha1 => SHA1.HashSizeInBytes,
+                SourceHashAlgorithm.Sha256 => SHA256.HashSizeInBytes,
+                _ => throw ExceptionUtilities.UnexpectedValue(algorithmId)
+            };
+
+            Span<byte> hashBuffer = stackalloc byte[size];
+            if (!algorithm.TryComputeHash(buffer, hashBuffer, out int bytesWritten))
             {
-                RoslynDebug.Assert(algorithm != null);
-                return ImmutableArray.Create(algorithm.ComputeHash(buffer, offset, count));
+                throw new InvalidOperationException();
             }
+
+            return ImmutableArray.Create(hashBuffer.Slice(0, bytesWritten));
+        }
+#endif
+
+        internal static ImmutableArray<byte> CalculateChecksum(byte[] buffer, int offset, int count, SourceHashAlgorithm algorithmId)
+        {
+#if NETCOREAPP
+            return CalculateChecksum(new ReadOnlySpan<byte>(buffer, offset, count), algorithmId);
+#else
+            using var algorithm = CryptographicHashProvider.GetAlgorithm(algorithmId);
+            var bytes = algorithm.ComputeHash(buffer, offset, count);
+
+            // ComputeHash return a freshly allocated array so it's safe to convert to ImmutableArray inline.
+            return ImmutableCollectionsMarshal.AsImmutableArray(bytes);
+#endif
         }
 
         internal static ImmutableArray<byte> CalculateChecksum(Stream stream, SourceHashAlgorithm algorithmId)
         {
-            using (var algorithm = CryptographicHashProvider.TryGetAlgorithm(algorithmId))
+            using var algorithm = CryptographicHashProvider.GetAlgorithm(algorithmId);
+            if (stream.CanSeek)
             {
-                RoslynDebug.Assert(algorithm != null);
-                if (stream.CanSeek)
+                stream.Seek(0, SeekOrigin.Begin);
+
+#if NETCOREAPP
```

Comment:
Similar comment; what makes this path faster over directly transmutting to an ImmutableArray?